src/jbi/services.py

```diff
@@ -127,6 +129,50 @@ def _all_jira_projects_visible(jira, actions: Actions) -> bool:
     return not missing_projects
 
 
+def _all_jira_projects_permissions(jira, actions: Actions):
+    all_projects_perms = {}
+    # Query permissions for all configured projects in parallel threads.
+    with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
+        futures_to_projects = {
+            executor.submit(
+                jira.get_permissions,
+                project_key=project_key,
+                permissions="CREATE_ISSUES,DELETE_ISSUES,EDIT_ISSUES",
```

Comment:
I think in the future we might need more permissions--so I thought it would make sense to convert it to a env var in the settings class. Let me know your thoughts

Reply:
I'm not sure it needs to be an env var. Will these permissions change based on whether we're in nonprod, prod, or some other specific environment? If so, then an env var makes sense to me. Otherwise, I think a constant is fine.

Reply:
It might change per-env if custom-actions use the jira-sdk in more advanced ways. Also, there are likely other permissions we already want to validate on (like ADD_COMMENTS).

```json
{'permissions': {'ADD_COMMENTS': {'key': 'ADD_COMMENTS', 'name': 'Add Comments', 'type': 'PROJECT', 'description': 'Ability to comment on issues.'}, 'ADMINISTER': {'key': 'ADMINISTER', 'name': 'Administer Jira', 'type': 'GLOBAL', 'description': 'Create and administer projects, issue types, fields, workflows, and schemes for all projects. Users with this permission can perform most administration tasks, except: managing users, importing data, and editing system email settings.'}, 'ADMINISTER_PROJECTS': {'key': 'ADMINISTER_PROJECTS', 'name': 'Administer Projects', 'type': 'PROJECT', 'description': 'Ability to administer a project in Jira.'}, 'ASSIGNABLE_USER': {'key': 'ASSIGNABLE_USER', 'name': 'Assignable User', 'type': 'PROJECT', 'description': 'Users with this permission may be assigned to issues.'}, 'ASSIGN_ISSUES': {'key': 'ASSIGN_ISSUES', 'name': 'Assign Issues', 'type': 'PROJECT', 'description': 'Ability to assign issues to other people.'}, 'BROWSE_PROJECTS': {'key': 'BROWSE_PROJECTS', 'name': 'Browse Projects', 'type': 'PROJECT', 'description': 'Ability to browse projects and the issues within them.'}, 'BULK_CHANGE': {'key': 'BULK_CHANGE', 'name': 'Make bulk changes', 'type': 'GLOBAL', 'description': 'Modify collections of issues at once. For example, resolve multiple issues in one step.'}, 'CLOSE_ISSUES': {'key': 'CLOSE_ISSUES', 'name': 'Close Issues', 'type': 'PROJECT', 'description': 'Ability to close issues. Often useful where your developers resolve issues, and a QA department closes them.'}, 'CREATE_ATTACHMENTS': {'key': 'CREATE_ATTACHMENTS', 'name': 'Create Attachments', 'type': 'PROJECT', 'description': 'Users with this permission may create attachments.'}, 'CREATE_ISSUES': {'key': 'CREATE_ISSUES', 'name': 'Create Issues', 'type': 'PROJECT', 'description': 'Ability to create issues.'}, 'CREATE_PROJECT': {'key': 'CREATE_PROJECT', 'name': 'Create team-managed projects', 'type': 'GLOBAL', 'description': "Create projects separate from shared configurations and schemes. Team-managed projects don't affect existing projects or shared configurations like workflows, fields or permissions. Only licensed users can create team-managed projects."}, 'CREATE_SHARED_OBJECTS': {'key': 'CREATE_SHARED_OBJECTS', 'name': 'Share dashboards and filters', 'type': 'GLOBAL', 'description': 'Share dashboards and filters with other users.'}, 'DELETE_ALL_ATTACHMENTS': {'key': 'DELETE_ALL_ATTACHMENTS', 'name': 'Delete All Attachments', 'type': 'PROJECT', 'description': 'Users with this permission may delete all attachments.'}, 'DELETE_ALL_COMMENTS': {'key': 'DELETE_ALL_COMMENTS', 'name': 'Delete All Comments', 'type': 'PROJECT', 'description': 'Ability to delete all comments made on issues.'}, 'DELETE_ALL_WORKLOGS': {'key': 'DELETE_ALL_WORKLOGS', 'name': 'Delete All Worklogs', 'type': 'PROJECT', 'description': 'Ability to delete all worklogs made on issues.'}, 'DELETE_ISSUES': {'key': 'DELETE_ISSUES', 'name': 'Delete Issues', 'type': 'PROJECT', 'description': 'Ability to delete issues.'}, 'DELETE_OWN_ATTACHMENTS': {'key': 'DELETE_OWN_ATTACHMENTS', 'name': 'Delete Own Attachments', 'type': 'PROJECT', 'description': 'Users with this permission may delete own attachments.'}, 'DELETE_OWN_COMMENTS': {'key': 'DELETE_OWN_COMMENTS', 'name': 'Delete Own Comments', 'type': 'PROJECT', 'description': 'Ability to delete own comments made on issues.'}, 'DELETE_OWN_WORKLOGS': {'key': 'DELETE_OWN_WORKLOGS', 'name': 'Delete Own Worklogs', 'type': 'PROJECT', 'description': 'Ability to delete own worklogs made on issues.'}, 'EDIT_ALL_COMMENTS': {'key': 'EDIT_ALL_COMMENTS', 'name': 'Edit All Comments', 'type': 'PROJECT', 'description': 'Ability to edit all comments made on issues.'}, 'EDIT_ALL_WORKLOGS': {'key': 'EDIT_ALL_WORKLOGS', 'name': 'Edit All Worklogs', 'type': 'PROJECT', 'description': 'Ability to edit all worklogs made on issues.'}, 'EDIT_ISSUES': {'key': 'EDIT_ISSUES', 'name': 'Edit Issues', 'type': 'PROJECT', 'description': 'Ability to edit issues.'}, 'EDIT_OWN_COMMENTS': {'key': 'EDIT_OWN_COMMENTS', 'name': 'Edit Own Comments', 'type': 'PROJECT', 'description': 'Ability to edit own comments made on issues.'}, 'EDIT_OWN_WORKLOGS': {'key': 'EDIT_OWN_WORKLOGS', 'name': 'Edit Own Worklogs', 'type': 'PROJECT', 'description': 'Ability to edit own worklogs made on issues.'}, 'LINK_ISSUES': {'key': 'LINK_ISSUES', 'name': 'Link Issues', 'type': 'PROJECT', 'description': 'Ability to link issues together and create linked issues. Only useful if issue linking is turned on.'}, 'MANAGE_GROUP_FILTER_SUBSCRIPTIONS': {'key': 'MANAGE_GROUP_FILTER_SUBSCRIPTIONS', 'name': 'Manage group filter subscriptions', 'type': 'GLOBAL', 'description': 'Create and delete group filter subscriptions.'}, 'MANAGE_SPRINTS_PERMISSION': {'key': 'MANAGE_SPRINTS_PERMISSION', 'name': 'Manage sprints', 'type': 'PROJECT', 'description': 'Ability to manage sprints.'}, 'MANAGE_WATCHERS': {'key': 'MANAGE_WATCHERS', 'name': 'Manage Watchers', 'type': 'PROJECT', 'description': 'Ability to manage the watchers of an issue.'}, 'MODIFY_REPORTER': {'key': 'MODIFY_REPORTER', 'name': 'Modify Reporter', 'type': 'PROJECT', 'description': 'Ability to modify the reporter when creating or editing an issue.'}, 'MOVE_ISSUES': {'key': 'MOVE_ISSUES', 'name': 'Move Issues', 'type': 'PROJECT', 'description': 'Ability to move issues between projects or between workflows of the same project (if applicable). Note the user can only move issues to a project they have the create permission for.'}, 'RESOLVE_ISSUES': {'key': 'RESOLVE_ISSUES', 'name': 'Resolve Issues', 'type': 'PROJECT', 'description': 'Ability to resolve and reopen issues. This includes the ability to set a fix version.'}, 'SCHEDULE_ISSUES': {'key': 'SCHEDULE_ISSUES', 'name': 'Schedule Issues', 'type': 'PROJECT', 'description': "Ability to view or edit an issue's due date."}, 'SERVICEDESK_AGENT': {'key': 'SERVICEDESK_AGENT', 'name': 'Service Project Agent', 'type': 'PROJECT', 'description': 'Allows users to interact with customers and access Jira Service Management features of a project.'}, 'SET_ISSUE_SECURITY': {'key': 'SET_ISSUE_SECURITY', 'name': 'Set Issue Security', 'type': 'PROJECT', 'description': 'Ability to set the level of security on an issue so that only people in that security level can see the issue.'}, 'SYSTEM_ADMIN': {'key': 'SYSTEM_ADMIN', 'name': 'Jira System Administrators', 'type': 'GLOBAL', 'description': 'Ability to perform all administration functions. There must be at least one group with this permission.'}, 'TRANSITION_ISSUES': {'key': 'TRANSITION_ISSUES', 'name': 'Transition Issues', 'type': 'PROJECT', 'description': 'Ability to transition issues.'}, 'USER_PICKER': {'key': 'USER_PICKER', 'name': 'Browse users and groups', 'type': 'GLOBAL', 'description': 'View and select users or groups from the user picker, and share issues. Users with this permission can see the names of all users and groups on your site.'}, 'VIEW_AGGREGATED_DATA': {'key': 'VIEW_AGGREGATED_DATA', 'name': 'View aggregated data', 'type': 'PROJECT', 'description': 'Users with this permission will have access to view combined and summarized project data, regardless of their individual permissions.'}, 'VIEW_DEV_TOOLS': {'key': 'VIEW_DEV_TOOLS', 'name': 'View Development Tools', 'type': 'PROJECT', 'description': 'Allows users in a software project to view development-related information on the issue, such as commits, reviews and build information.'}, 'VIEW_READONLY_WORKFLOW': {'key': 'VIEW_READONLY_WORKFLOW', 'name': 'View Read-Only Workflow', 'type': 'PROJECT', 'description': 'Users with this permission may view a read-only version of a workflow.'}, 'VIEW_VOTERS_AND_WATCHERS': {'key': 'VIEW_VOTERS_AND_WATCHERS', 'name': 'View Voters and Watchers', 'type': 'PROJECT', 'description': 'Ability to view the voters and watchers of an issue.'}, 'WORK_ON_ISSUES': {'key': 'WORK_ON_ISSUES', 'name': 'Work On Issues', 'type': 'PROJECT', 'description': 'Ability to log work done against an issue. Only useful if Time Tracking is turned on.'}}}
```

Reply:
I believe the set of required permissions is more coupled to the actions implementation than to the env.

What if the permissions were class attributes on the action instead? With this design, the heartbeat would check that all projects have all permissions, but would iterate configured projects and check that the permissions required by the action being used are in place.

Reply:
I opened #177 for this